extension/ackextension/inmemory.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ackextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/ackextension"
+
+import (
+	"context"
+	"sync/atomic"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"go.opentelemetry.io/collector/component"
+)
+
+// InMemoryAckExtension is the in-memory implementation of the AckExtension
+// When MaxNumPartition is reached, the acks associated with the least recently used partition are evicted.
+// When MaxNumPendingAcksPerPartition is reached, the least recently used ack is evicted
+type InMemoryAckExtension struct {
+	partitionMap                  *lru.Cache[string, *ackPartition]
+	maxNumPendingAcksPerPartition uint64
+}
+
+func newInMemoryAckExtension(conf *Config) *InMemoryAckExtension {
+	cache, _ := lru.New[string, *ackPartition](int(conf.MaxNumPartition))
+	return &InMemoryAckExtension{
+		partitionMap:                  cache,
+		maxNumPendingAcksPerPartition: conf.MaxNumPendingAcksPerPartition,
+	}
+}
+
+type ackPartition struct {
+	id     atomic.Uint64
+	ackMap *lru.Cache[uint64, bool]
+}
+
+func newAckPartition(maxPendingAcks uint64) *ackPartition {
+	cache, _ := lru.New[uint64, bool](int(maxPendingAcks))
+	return &ackPartition{
+		ackMap: cache,
+	}
+}
+
+func (as *ackPartition) nextAck() uint64 {
+	id := as.id.Add(1)
+	as.ackMap.Add(id, false)
+	return id
+}
+
+func (as *ackPartition) ack(key uint64) {
+	if _, ok := as.ackMap.Get(key); ok {
+		as.ackMap.Add(key, true)
+	}
+}
+
+func (as *ackPartition) computeAcks(ackIDs []uint64) map[uint64]bool {
+	result := make(map[uint64]bool, len(ackIDs))
+	for _, val := range ackIDs {
+		if isAcked, ok := as.ackMap.Get(val); ok && isAcked {
+			result[val] = true
+			as.ackMap.Remove(val)
```

Comment:
Why we remove the ack on the query? I thought `QueryAcks` is idempotent and acks expire indenedenly. This behavior sounds a bit surprising. If really necessary, I think it should be better documented in `QueryAcks`. cc @splunkericl

Reply:
This behavior is described in https://docs.splunk.com/Documentation/Splunk/9.1.0/Data/AboutHECIDXAck#:~:text=Because%20a%20false,as%20true.

Reply:
Ok I believe this behavior can be configurable in future if needed